lib/delegate/index.js

```diff
@@ -0,0 +1,149 @@
+      kind: Kind.FIELD,
+      name: { kind: Kind.NAME, value: '__typename' }
+    });
+  }
+
+  // extract the arguments from the calling resolver
+  let newArgs = info.fieldNodes.reduce((acc, fieldNode) => {
+    if (fieldNode.arguments) {
+      return acc.concat(fieldNode.arguments);
+    }
+  }, []);
+
+  const targetRootFieldNode = {
+    kind: Kind.FIELD,
+    arguments: newArgs,
+    name: { kind: Kind.NAME, value: targetRootField },
+    selectionSet: { kind: Kind.SELECTION_SET, selections }
+  };
+
+  const operationDefinition = {
+    kind: Kind.OPERATION_DEFINITION,
+    operation: info.operation.operation,
+    selectionSet: { kind: Kind.SELECTION_SET, selections: [targetRootFieldNode]},
+    variableDefinitions: info.operation.variableDefinitions
+  }
+
+  const definitions = [operationDefinition]
+  for (const [, fragmentDefinition] of Object.entries(info.fragments)) {
+    definitions.push(fragmentDefinition);
+  }
+  return {
+    kind: Kind.DOCUMENT,
+    definitions
+  }
+}
+
+/**
+ * executes (delegates) a graphql operation on the input component's schema
+ * @param {GraphQLComponent} component - the component to delegate execution to
+ * @param {object} options - an options object to customize the delegated 
+ * operation
+ * @param {GraphQLResolveInfo} options.info - the info object from the calling resolver
+ * @param {object} options.contextValue - the context object from the calling 
+ * resolver
+ * @param {string} [options.targetRootField] - the name of the root type field 
+ * the delegated operation will execute. Defaults to the field name of the 
+ * calling resolver
+ * @param {string} [options.subPath] - a dot separated string to limit the 
+ * delegated selection set to a given path in the calling resolver's return type
+ * @returns the result of the delegated operation to the targetRootField with
+ * any errors merged into the result at their given path
+ */
+const delegateToComponent = async function (component, options) { 
+  let {
+    subPath,
+    contextValue,
+    info,
+    targetRootField
+  } = options;
+
+  if (!contextValue) {
+    throw new Error('delegateToComponent requires the contextValue from the calling resolver');
+  }
+
+  if (!info) {
+    throw new Error('delegateToComponent requires the info object from the calling resolver');
+  }
+
+  // default the target root field to be the name of the calling resolver
+  if (!targetRootField) {
+    targetRootField = info.fieldName;
+  }
+
+  const document = createSubOperationDocument(targetRootField, subPath, info);
+
+  debug(`delegating ${print(document)} to ${component.name}`);
+
+  const { data = {}, errors = [] } = await execute({
+    document, 
+    schema: component.schema, 
+    rootValue: info.rootValue, 
+    contextValue, 
+    variableValues: info.variableValues
+  });
+  
+  if (errors.length > 0) {
+    for (const error of errors) {
+      deepSet(data, error.path, error);
+    }
+  }
+
+  return data[targetRootField];
```

Comment:
Manas pointed me to a problem he was seeing that led me to find a bug here ... when the sub document's return type is non nullable, graphql can null out the entire response all they way up through `result.data` ie.

if I'm executing:
```graphql
query {
  test {
    value
  }
}
```
on 

```graphql
type Query {
  test: Test!
}

type Test {
  value: Boolean
}
```
and 

```javascript
const resolvers = {
  Query: {
    test() { throw new Error('error') }
  }
}
```

```javascript
 const { data = {}, errors = [] } = graphq.execute(...);
```
`data` will be `null` and not `{ test: null }` .... destructing defaults dont apply to `null`, only undefined, so the following deepSet of the errors at path doesn't get applied and the return statement here throws an error when it tries to key into data instead properly returning the error from the resolver